gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			notify_flow_flush(ip_prefix_fib,
+				&gk_conf->instances[i]);
+		}
+
+		/*
+		 * Wait until all the GK instances
+		 * finish to flush their flow tables.
+		 */
+		while (rte_atomic16_read(&ip_prefix_fib->num_updated_instances)
+			!= gk_conf->num_lcores);
+	}
+
+	/* Reset the fields of the deleted FIB entry. */
+	ip_prefix_fib->action = GK_FIB_MAX;
+	ip_prefix_fib->ref_cnt = 0;
+	rte_atomic16_init(&ip_prefix_fib->num_updated_instances);
+	memset(&ip_prefix_fib->u, 0, sizeof(ip_prefix_fib->u));
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: delete a fib entry [ip prefix = %s]\n", ip_prefix);
+
+	return 0;
+}
+
+/*
+ * Initialize a gateway FIB entry.
+ * @gateway the gateway address informaiton.
+ * @ip_prefix the IP prefix,
+ * for which the gateway is responsible.
+ */
+static struct gk_fib *
+init_gateway_fib(const char *ip_prefix, enum gk_fib_action action,
+	const char *gateway, struct gk_config *gk_conf)
+{
+	int ret;
+	int fib_id = -1;
+	int ip_prefix_len;
+	struct ipaddr gw_addr;
+	struct ipaddr ip_prefix_addr;
+	struct ether_cache *eth_cache;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_fib *neigh_fib = NULL;
+	struct gk_fib *ip_prefix_fib = NULL;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gatekeeper_if *iface;
+
+	/* Initialize the fib entry for the gateway. */
+	ret = convert_str_to_ip(gateway, &gw_addr);
+	if (ret < 0)
+		return NULL;
+
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	if (gw_addr.proto != ip_prefix_addr.proto) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since the gateway and its responsible grantors have different IP versions.\n");
+		return NULL;
+	}
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (action == GK_FWD_GATEWAY_BACK_NET)
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since it has invalid action %d.\n",
+			action);
+		return NULL;
+	}
+
+	/* Find the FIB entry for the gateway IP address. */
+	if (gw_addr.proto == ETHER_TYPE_IPv4 &&
+			ipv4_if_configured(iface)) {
+		fib_id = lpm_lookup_ipv4(ltbl->lpm, gw_addr.ip.v4.s_addr);
+		/*
+		 * Invalid gateway entry, since at least we should
+		 * obtain the FIB entry for the neighbor table.
+		 */
+		if (fib_id < 0)
+			return NULL;
+
+		neigh_fib = &ltbl->fib_tbl[fib_id];
+
+		/*
+		 * Invalid gateway entry, since the neighbor entry
+		 * and the gateway entry should be in the same network.
+		 */
+		if ((action == GK_FWD_GATEWAY_FRONT_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_FRONT_NET
+				&& neigh_fib->action !=
+				GK_FWD_GATEWAY_FRONT_NET)
+				|| (action == GK_FWD_GATEWAY_BACK_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_BACK_NET &&
+				neigh_fib->action != GK_FWD_GATEWAY_BACK_NET)) {
+			return NULL;
+		}
+
+		/* The gateway FIB entry is existing. */
```

Comment:
You can just say "The gateway FIB entry exists." Look for similar cases of "is existing" in this patch as well.